### ISSUE TITLE
server: TLS error from health check doesn't close connections

### ISSUE THREAD

**Issue by mjibson** [cockroachdb/cockroach#19171]
Version: CockroachDB CCL v1.1-beta.20170921 (linux amd64, built 2017/09/21 15:48:53, go1.8.3)

A customer had a cluster on GCE where all nodes were using 200% CPU, not printing anything to logs, and not responding to any network request. Even `curl http://localhost:8080` resulted in a timeout. There were errors in the logs of the form `E171008 14:45:05.565066 32651905 server.go:2751  http: TLS handshake error from 130.211.0.226:62246: EOF` about once a second, which was a health check by the google load balancer on port 8080. The cluster was running in secure mode. `strace -f` showed many new threads all erroring with "too many open files". lsof and netstat showed that the process had thousands of tcp connections in the CLOSE_WAIT state.

It seems that the health check was causing some kind of connection error that was not being cleaned up correctly. These piled up eventually causing the server to enter an infinite loop.

Upon `pkill -sigabrt cockroach`, stack traces were printed to the log. Many were of the form:

```
goroutine 35010048 [semacquire, 2335 minutes]:
sync.runtime_SemacquireMutex(0xc42042b904)
	/usr/local/go/src/runtime/sema.go:62 +0x34
sync.(*Mutex).Lock(0xc42042b900)
	/usr/local/go/src/sync/mutex.go:87 +0x9d
log.(*Logger).Output(0xc42042b900, 0x2, 0xc430aeb300, 0x37, 0x0, 0x0)
	/usr/local/go/src/log/log.go:149 +0x7a
log.(*Logger).Printf(0xc42042b900, 0x1d2decf, 0x25, 0xc44a410ef0, 0x2, 0x2)
	/usr/local/go/src/log/log.go:175 +0x7e
net/http.(*Server).logf(0xc420280630, 0x1d2decf, 0x25, 0xc44a410ef0, 0x2, 0x2)
	/usr/local/go/src/net/http/server.go:2751 +0x6a
net/http.(*conn).serve(0xc42a7e2500, 0x2aba2c0, 0xc4201d42a0)
	/usr/local/go/src/net/http/server.go:1738 +0x226
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:2668 +0x2ce
```

There are at least two bugs here:

1. The connections should correctly close.
2. We need to better handle the case when we are out of file descriptors.

Current pending followup questions:

1. What was the exact health check? i.e., TCP on 8080, HTTP/HTTPS on 8080/health?
2. What did Google report about the health check (healthy/not health)?

@bdarnell @a-robinson can one of you triage this and assign to someone for investigation?

**Comment by mjibson** [comment]
Responses:

1. The health check was TCP on 8080.
2. Google's health checker was succeeding.

**Comment by cuongdo** [comment]
@a-robinson could you take a closer look?

**Comment by a-robinson** [comment]
@mjibson do you still have any of the terminal output from our debug session? Reproducing this apparently isn't as simple as setting up a secure cluster on gce with a tcp health check configured against port 8080, so I'm wondering if all the stuck CLOSE_WAIT connections may have actually caused by something else, like the repeated failed DNS lookups from trying to connect to the nodes that weren't around anymore.

**Comment by mjibson** [comment]
No, I don't have terminal output. But I may still have SSH access and the bug may still be manifesting. I will check tomorrow at the office.

**Comment by bdarnell** [comment]
That stack is waiting to acquire a mutex in the standard `log` package. I think this mutex is causing the `CLOSE_WAIT` connections rather than the other way around. Do you have a complete stack dump? There should be some thread that's doing something in the `cockroach/pkg/util/log` package. My guess is that there's a deadlock in a `String()` method called on a rarely-reached error path.

**Comment by nvanbenschoten** [comment]
The only candidate in the standard `log` package I see that could create a deadlock is `Logger.Output`, which calls out to the provided `io.Writer`. We supply this with a `cockroach/pkg/util/log.logBridge`, so if this guess is correct then I'd expect to see `logBridge.Write` somewhere in the goroutine dump.

@mjibson are you still in contact with the customer, or if not, could you point me at them? The
full stack dump would be able to confirm this suspicion.

**Comment by bdarnell** [comment]
OK, we've gotten a full goroutine dump and it looks like it was a combination of running out of a disk error (probably running out of space) and the bug fixed in #17871 (which will be included in version 1.1.1). The log shows a ton of goroutines in the middle of log.Fatal calls, which is where we'd expect to crash if the disk fills up.

```
1: semacquire [1352 minutes] [Created by stop.(*Stopper).RunWorker @ stopper.go:199]
    sync       sema.go:62                   runtime_SemacquireMutex(#106)
    sync       mutex.go:87                  (*Mutex).Lock(#105)
    syncutil   mutex_sync.go:32             (*Mutex).Lock(#105)
    log        clog.go:728                  (*loggingT).outputLogEntry(#104, #124, #64, 0x10, 0xdf2, #14411, 0x46)
    log        structured.go:162            addStructured(#45822, #539, #124, 0x2, 0, 0, #42486, 0x1, 0x1)
    log        log.go:54                    logDepth(#45822, #539, 0x1, 0x4, 0, 0, #42486, 0x1, 0x1)
    log        log.go:157                   Fatal(#45822, #539, #42486, 0x1, 0x1)
    storage    store.go:3570                (*Store).processReady(#831, #45822, #539, 0x3e1)
    storage    scheduler.go:216             (*raftScheduler).worker(#322, #45822, #539)
    storage    scheduler.go:167             (*raftScheduler).Start.func2(#45822, #539)
    stop       stopper.go:198               (*Stopper).RunWorker.func1(#242, #412, #241)
```

The TLS errors look like just a red herring. We still have #19202 to try and mitigate file descriptor leaks, although that wouldn't have helped much here, and we have #8473 and #8474 to handle out-of-disk-space errors more gracefully.

**Comment by nvanbenschoten** [comment]
@bdarnell just beat me!

The problematic goroutine that #17871 will now prevent was:
```
goroutine 247 [semacquire, 2539 minutes]:
sync.runtime_SemacquireMutex(0x2f7e228)
        /usr/local/go/src/runtime/sema.go:62 +0x34
sync.(*Mutex).Lock(0x2f7e224)
        /usr/local/go/src/sync/mutex.go:87 +0x9d
github.com/cockroachdb/cockroach/pkg/util/syncutil.(*Mutex).Lock(0x2f7e224)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/syncutil/mutex_sync.go:32 +0x2d
github.com/cockroachdb/cockroach/pkg/util/log.(*loggingT).outputLogEntry(0x2f7e200, 0xc400000003, 0x27c00f3, 0x1b, 0x72, 0xc423c4f880, 0x1a)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:728 +0x12b
github.com/cockroachdb/cockroach/pkg/util/log.addStructured(0x7f968d3a7168, 0xc420afe360, 0x3, 0x2, 0x0, 0x0, 0xc442f89768, 0x1, 0x1)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/structured.go:162 +0x48f
github.com/cockroachdb/cockroach/pkg/util/log.logDepth(0x7f968d3a7168, 0xc420afe360, 0x1, 0xc400000003, 0x0, 0x0, 0xc442f89768, 0x1, 0x1)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/log.go:54 +0x8c
github.com/cockroachdb/cockroach/pkg/util/log.Shout(0x7f968d3a7168, 0xc420afe360, 0x3, 0xc435249768, 0x1, 0x1)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/log.go:60 +0x9a
github.com/cockroachdb/cockroach/pkg/util/log.ReportPanic(0x7f968d3a7168, 0xc420afe360, 0xc420433300, 0x1ae86c0, 0xc434385530, 0x1)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/crash_reporting.go:114 +0xe1
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc4205ddb00, 0x7f968d3a7168, 0xc420afe360)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:181 +0xee
panic(0x1ae86c0, 0xc434385530)
        /usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/util/log.(*loggingT).outputToStderr(0x2f7e200, 0x4, 0x14ebc339e9ba2893, 0xf7, 0x27b10bd, 0x10, 0xbd, 0xc439782900, 0x88, 0xc42d9dd800, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:788 +0x11a
github.com/cockroachdb/cockroach/pkg/util/log.(*loggingT).outputLogEntry(0x2f7e200, 0xc400000004, 0x27b10bd, 0x10, 0xbd, 0xc439782900, 0x88)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:749 +0x227
github.com/cockroachdb/cockroach/pkg/util/log.addStructured(0x7f968d3a7168, 0xc4203b1980, 0x4, 0x2, 0x0, 0x0, 0xc442f89cf0, 0x1, 0x1)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/structured.go:162 +0x48f
github.com/cockroachdb/cockroach/pkg/util/log.logDepth(0x7f968d3a7168, 0xc4203b1980, 0x1, 0x4, 0x0, 0x0, 0xc442f89cf0, 0x1, 0x1)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/log.go:54 +0x8c
github.com/cockroachdb/cockroach/pkg/util/log.Fatal(0x7f968d3a7168, 0xc4203b1980, 0xc442f89cf0, 0x1, 0x1)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/log.go:157 +0x76
```
